rllib/core/models/base.py

```diff
@@ -410,32 +408,26 @@ def get_initial_state(self):
             }
 
     @override(Model)
-    def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
+    def _forward(self, inputs: dict, **kwargs) -> dict:
         if self.config.shared:
             outs = self.encoder(inputs, **kwargs)
-            return NestedDict(
-                {
-                    ENCODER_OUT: {ACTOR: outs[ENCODER_OUT], CRITIC: outs[ENCODER_OUT]},
-                    STATE_OUT: outs[STATE_OUT],
-                }
-            )
+            return {
+                ENCODER_OUT: {ACTOR: outs[ENCODER_OUT], CRITIC: outs[ENCODER_OUT]},
+                STATE_OUT: outs[STATE_OUT],
+            }
         else:
-            actor_inputs = NestedDict({**inputs})
-            # , **{STATE_IN: inputs[STATE_IN][ACTOR]}})
-            critic_inputs = NestedDict(
-                {**inputs}  # , **{STATE_IN: inputs[STATE_IN][CRITIC]}}
-            )
+            actor_inputs = inputs  # , **{STATE_IN: inputs[STATE_IN][ACTOR]}})
```

Comment:
Ah, yes, here we are already skipping the STATE key. Let's cleanly remove this everywhere some Encoder sub-class defines its in- and output-states (except for the LSTM/GRUEncoders, which should have these, of course).

Reply:
This is not super straight forward and a reason why I wanted to move it to another PR.
If you remove this here, encoder and decoder get the same inputs dict. Since they don't have internal knowledge of whether they are an actor or critic model, they would not be able to know which state they should use.
Therefore, only one of them could be stateful.
If we don't do this here, the separation of inputs for actor and critic leaks into the RL Module forward methods.

@kouroshHakha @sven1977 More input here, please.
I propose that in `get_initial_state()`, we check if both initial states are None.
If so, we don't pass around states.
If there are one or two initial states states, we create some logic to separate them in the ActorCriticEncoder so that this maintains transparent to RLModule and the encapsulated encoder models

Reply:
I agree that this is a complex change and let's move it to a separate [PR](https://github.com/ray-project/ray/pull/34911) (including the discussion). I remember that we went through several iterations of this with @ArturNiederfahrenhorst, but I forget the concrete examples of intricacies that  justified using this pattern. In that PR, @ArturNiederfahrenhorst, Can you give a few concrete examples that required us to carry STATE_IN and STATE_OUT keys around in the base class, I think we need them documented in the PR to avoid repetition and can drive the discussion from there.